### ISSUE TITLE
SSL module returns incorrect error codes on Windows

### ISSUE THREAD

**Issue by zooba** [python/cpython#75581]
BPO | [31400](https://bugs.python.org/issue31400)
--- | :---
Nosy | @tiran, @alex, @zooba, @dstufft
PRs | <li>python/cpython#3463</li><li>python/cpython#3466</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/zooba'
closed_at = <Date 2017-09-11.10:25:24.698>
created_at = <Date 2017-09-08.21:16:30.722>
labels = ['expert-SSL', 'type-bug', '3.7']
title = 'SSL module returns incorrect error codes on Windows'
updated_at = <Date 2017-09-11.10:25:24.696>
user = 'https://github.com/zooba'
```

bugs.python.org fields:
```python
activity = <Date 2017-09-11.10:25:24.696>
actor = 'christian.heimes'
assignee = 'steve.dower'
closed = True
closed_date = <Date 2017-09-11.10:25:24.698>
closer = 'christian.heimes'
components = ['SSL']
creation = <Date 2017-09-08.21:16:30.722>
creator = 'steve.dower'
dependencies = []
files = []
hgrepos = []
issue_num = 31400
keywords = ['patch']
message_count = 4.0
messages = ['301733', '301743', '301775', '301870']
nosy_count = 5.0
nosy_names = ['janssen', 'christian.heimes', 'alex', 'steve.dower', 'dstufft']
pr_nums = ['3463', '3466']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue31400'
versions = ['Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by zooba** [comment]
Because we wait so long between errors occurring and being reported, information about the Windows error code (WSAGetLastError) has already been lost.

We should capture all the info sooner.

**Comment by tiran** [comment]
New changeset e6eb48c10dc389d1d70657593de6a6cb3087d3d1 by Christian Heimes (Steve Dower) in branch 'master':
bpo-31400: Improve SSL error handling on Windows (bpo-3463)
https://github.com/python/cpython/commit/e6eb48c10dc389d1d70657593de6a6cb3087d3d1

**Comment by tiran** [comment]
New changeset 16f16dbd0ed02cd1a7b270eb6dd80d9bd179902e by Christian Heimes in branch '3.6':
[3.6] bpo-31400: Improve SSL error handling on Windows (GH-3463) (bpo-3466)
https://github.com/python/cpython/commit/16f16dbd0ed02cd1a7b270eb6dd80d9bd179902e

**Comment by tiran** [comment]
Patch has landed in 3.6 and master, thanks!
